File: src/packages/sdk/src/runtime/eval.ts
/**
 * EvalGate defineEval() DSL - Layer 1 Foundation
 *
 * The core DSL function for defining behavioral specifications.
 * Uses content-addressable identity with AST position for stability.
 */

import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";
import { getActiveRuntime } from "./registry";
import type {
	DefineEvalFunction,
	EvalContext,
	EvalExecutor,
	EvalResult,
	EvalSpec,
	SpecConfig,
	SpecOptions,
} from "./types";
import { SpecRegistrationError } from "./types";

/**
 * Extract AST position from call stack
 * This provides stable identity that survives renames but changes when logic moves
 */
function getCallerPosition(): {
	line: number;
	column: number;
	filePath: string;
} {
	const stack = new Error().stack;
	if (!stack) {
		throw new SpecRegistrationError("Unable to determine caller position");
	}

	// Parse stack trace to find the caller
	const lines = stack.split("\n");

	// Skip current function and find the actual caller
	for (let i = 3; i < lines.length; i++) {
		const line = lines[i];
		if (
			!line ||
			line.includes("node_modules") ||
			line.includes("internal/modules")
		) {
			continue;
		}

		// Extract file path, line, and column
		const match = line.match(/at\s+.*?\((.*?):(\d+):(\d+)\)/);
		if (match) {
			const [, filePath, lineNum, colNum] = match;
			return {
				filePath: path.resolve(filePath),
				line: parseInt(lineNum, 10),
				column: parseInt(colNum, 10),
			};
		}

		// Alternative format for some environments
		const altMatch = line.match(/at\s+(.*?):(\d+):(\d+)/);
		if (altMatch) {
			const [, filePath, lineNum, colNum] = altMatch;
			return {
				filePath: path.resolve(filePath),
				line: parseInt(lineNum, 10),
				column: parseInt(colNum, 10),
			};
		}
	}

	throw new SpecRegistrationError(
		"Unable to parse caller position from stack trace",
	);
}

/**
 * Generate content-addressable specification ID
 */
function generateSpecId(
	namespace: string,
	filePath: string,
	name: string,
	position: { line: number; column: number },
): string {
	// Canonicalize path: relative to project root with POSIX separators
	const projectRoot = process.cwd();
	const relativePath = path.relative(projectRoot, filePath);
	const canonicalPath = relativePath.split(path.sep).join("/"); // Force POSIX separators

	const components = [
		namespace,
		canonicalPath,
		name,
		`${position.line}:${position.column}`,
	];

	const content = components.join("|");
	return crypto.createHash("sha256").update(content).digest("hex").slice(0, 20);
}

/**
 * Validate specification name
 */
function validateSpecName(name: string): void {
	if (!name || typeof name !== "string") {
		throw new SpecRegistrationError(
			"Specification name must be a non-empty string",
		);
	}

	if (name.trim() === "") {
		throw new SpecRegistrationError("Specification name cannot be empty");
	}

	if (name.length > 100) {
		throw new SpecRegistrationError(
			"Specification name must be 100 characters or less",
		);
	}

	// Check for invalid characters
	if (!/^[a-zA-Z0-9\s\-_]+$/.test(name)) {
		throw new SpecRegistrationError(
			"Specification name can only contain letters, numbers, spaces, hyphens, and underscores",
		);
	}
}

/**
 * Validate executor function
 */
function validateExecutor(executor: EvalExecutor): void {
	if (typeof executor !== "function") {
		throw new SpecRegistrationError("Executor must be a function");
	}

	// Check function length (should accept context parameter)
	if (executor.length > 1) {
		throw new SpecRegistrationError(
			"Executor should accept exactly one parameter (context)",
		);
	}
}

/**
 * Create specification configuration from parameters
 */
function createSpecConfig(
	nameOrConfig: string | SpecConfig,
	executor?: EvalExecutor,
	options?: SpecOptions,
): SpecConfig {
	if (typeof nameOrConfig === "string") {
		// defineEval(name, executor, options) form
		if (!executor) {
			throw new SpecRegistrationError(
				"Executor function is required when using name parameter",
			);
		}

		return {
			name: nameOrConfig,
			executor,
			...options,
		};
	} else {
		// defineEval(config) form
		return nameOrConfig;
	}
}

/**
 * Core defineEval function implementation
 */
function defineEvalWithMode(
	mode: "normal" | "skip" | "only",
	nameOrConfig: string | SpecConfig,
	executor?: EvalExecutor,
	options?: SpecOptions,
): void {
	// Get caller position for identity
	const callerPosition = getCallerPosition();

	// Create specification configuration
	const config = createSpecConfig(nameOrConfig, executor, options);

	// Validate configuration
	validateSpecName(config.name);
	validateExecutor(config.executor);

	// Get active runtime
	const runtime = getActiveRuntime();

	// Generate specification ID
	const specId = generateSpecId(
		runtime.namespace,
		callerPosition.filePath,
		config.name,
		callerPosition,
	);

	// Create specification
	const spec: EvalSpec = {
		id: specId,
		name: config.name,
		filePath: callerPosition.filePath,
		position: callerPosition,
		description: config.description,
		tags: config.tags,
		executor: config.executor,
		metadata: config.metadata,
		config: {
			timeout: config.timeout,
			retries: config.retries,
			budget: config.budget,
			model: config.model,
		},
		mode,
	};

	// Register specification
	runtime.register(spec);
}

function defineEvalImpl<_TInput = string>(
	nameOrConfig: string | SpecConfig,
	executor?: EvalExecutor,
	options?: SpecOptions,
): void {
	defineEvalWithMode("normal", nameOrConfig, executor, options);
}

function defineEvalSkipImpl<_TInput = string>(
	nameOrConfig: string | SpecConfig,
	executor?: EvalExecutor,
	options?: SpecOptions,
): void {
	defineEvalWithMode("skip", nameOrConfig, executor, options);
}

function defineEvalOnlyImpl<_TInput = string>(
	nameOrConfig: string | SpecConfig,
	executor?: EvalExecutor,
	options?: SpecOptions,
): void {
	defineEvalWithMode("only", nameOrConfig, executor, options);
}

/**
 * Export the defineEval function with proper typing
 * This is the main DSL entry point
 */
export const defineEval: DefineEvalFunction =
	defineEvalImpl as DefineEvalFunction;

// Attach .skip and .only modifiers (vitest/jest convention)
defineEval.skip = defineEvalSkipImpl as DefineEvalFunction;
defineEval.only = defineEvalOnlyImpl as DefineEvalFunction;

/**
 * Parse a JSONL file into an array of row objects.
 * Each line must be a valid JSON object; blank lines are skipped.
 */
function parseJsonl(content: string): Record<string, unknown>[] {
	return content
		.split("\n")
		.map((line) => line.trim())
		.filter((line) => line.length > 0)
		.map((line, i) => {
			try {
				return JSON.parse(line) as Record<string, unknown>;
			} catch {
				throw new SpecRegistrationError(
					`Invalid JSON on line ${i + 1} of dataset`,
				);
			}
		});
}

/**
 * Parse a simple CSV file into an array of row objects.
 * First line is treated as headers. Values are unquoted strings.
 * For complex CSV (quoted fields, escapes), use a dedicated library.
 */
function parseCsv(content: string): Record<string, unknown>[] {
	const lines = content
		.split("\n")
		.map((l) => l.trim())
		.filter((l) => l.length > 0);
	if (lines.length < 2) return [];
	const headers = lines[0].split(",").map((h) => h.trim());
	return lines.slice(1).map((line) => {
		const values = line.split(",").map((v) => v.trim());
		const row: Record<string, unknown> = {};
		for (let i = 0; i < headers.length; i++) {
			row[headers[i]] = values[i] ?? "";
		}
		return row;
	});
}

/**
 * Load a JSONL or CSV dataset and register one spec per row.
 */
function fromDatasetImpl<TRow extends Record<string, unknown>>(
	name: string,
	datasetPath: string,
	executor: (context: EvalContext & { input: TRow }) => Promise<EvalResult>,
	options?: SpecOptions,
): void {
	const resolvedPath = path.isAbsolute(datasetPath)
		? datasetPath
		: path.resolve(process.cwd(), datasetPath);

	if (!fs.existsSync(resolvedPath)) {
		throw new SpecRegistrationError(
			`Dataset file not found: ${resolvedPath}`,
		);
	}

	const content = fs.readFileSync(resolvedPath, "utf8");
	const ext = path.extname(resolvedPath).toLowerCase();

	let rows: Record<string, unknown>[];
	if (ext === ".jsonl" || ext === ".ndjson") {
		rows = parseJsonl(content);
	} else if (ext === ".csv") {
		rows = parseCsv(content);
	} else if (ext === ".json") {
		const parsed = JSON.parse(content);
		rows = Array.isArray(parsed) ? parsed : [parsed];
	} else {
		throw new SpecRegistrationError(
			`Unsupported dataset format: ${ext}. Use .jsonl, .ndjson, .csv, or .json`,
		);
	}

	if (rows.length === 0) {
		throw new SpecRegistrationError(
			`Dataset is empty: ${resolvedPath}`,
		);
	}

	for (let i = 0; i < rows.length; i++) {
		const row = rows[i] as TRow;
		const specName = `${name} - row ${i + 1}`;
		const wrappedExecutor: EvalExecutor = (context: EvalContext) =>
			executor({ ...context, input: row as unknown as string & TRow });
		defineEvalWithMode("normal", specName, wrappedExecutor, {
			...options,
			metadata: {
				...options?.metadata,
				datasetPath: resolvedPath,
				datasetRow: i + 1,
			},
		});
	}
}

defineEval.fromDataset = fromDatasetImpl as DefineEvalFunction["fromDataset"];

/**
 * Filter a list of specs according to skip/only semantics:
 * - If any spec has mode === "only", return only those specs
 * - Otherwise, return all specs except those with mode === "skip"
 */
export function getFilteredSpecs(specs: EvalSpec[]): EvalSpec[] {
	const onlySpecs = specs.filter((s) => s.mode === "only");
	if (onlySpecs.length > 0) {
		return onlySpecs;
	}
	return specs.filter((s) => s.mode !== "skip");
}

/**
 * Convenience export for evalai.test() alias (backward compatibility)
 * Provides alternative naming that matches the original roadmap vision
 */
export const evalai = {
	test: defineEval,
};

/**
 * Suite definition for grouping related specifications.
 * Accepts both a positional form and an object form:
 *
 * @example Positional form:
 * defineSuite('My Suite', [() => defineEval('spec 1', executor), ...])
 *
 * @example Object form:
 * defineSuite({ name: 'My Suite', specs: [() => defineEval('spec 1', executor), ...] })
 */
export function defineSuite(
	nameOrConfig: string | { name: string; specs: (() => void)[] },
	specsArg?: (() => void)[],
): void {
	const specFns =
		typeof nameOrConfig === "string"
			? (specsArg ?? [])
			: (nameOrConfig.specs ?? []);
	// Execute each spec function to register its defineEval calls
	// In Layer 3, this will also build the dependency graph
	for (const specFn of specFns) {
		specFn();
	}
}

/**
 * Helper function to create specification contexts
 * Useful for testing and manual execution
 */
export function createContext<TInput = string>(
	input: TInput,
	metadata?: Record<string, unknown>,
	options?: EvalContext["options"],
): EvalContext & { input: TInput } {
	return {
		input: input as string & TInput,
		metadata,
		options,
	};
}

/**
 * Helper function to create specification results
 * Provides a convenient builder pattern for common result patterns
 */
export function createResult(config: {
	pass: boolean;
	score: number;
	assertions?: EvalResult["assertions"];
	metadata?: Record<string, unknown>;
	error?: string;
	output?: string;
	durationMs?: number;
	tokens?: number;
}): EvalResult {
	return {
		pass: config.pass,
		score: Math.max(0, Math.min(100, config.score)), // Clamp to 0-100
		assertions: config.assertions,
		metadata: config.metadata,
		error: config.error,
		output: config.output,
		durationMs: config.durationMs,
		tokens: config.tokens,
	};
}

/**
 * Default export for convenience
 */
export default defineEval;

File: src/packages/sdk-python/src/evalgate_sdk/runtime/eval.py
"""defineEval DSL — the primary API for declaring evaluation specs."""

from __future__ import annotations

import csv
import hashlib
import inspect
import io
import json
import os
import re
from typing import Any, Callable, Literal

from evalgate_sdk.runtime.registry import get_active_runtime
from evalgate_sdk.runtime.types import (
    EvalContext,
    EvalResult,
    EvalSpec,
    SpecConfig,
    SpecOptions,
    SpecRegistrationError,
)

_NAME_PATTERN = re.compile(r"^[\w\s\-]{1,100}$")


def _generate_spec_id(name: str, file_path: str | None = None) -> str:
    """Generate a content-addressable spec ID."""
    source = name
    if file_path is None:
        frame = inspect.stack()
        for f in frame[1:]:
            if "evalgate_sdk" not in f.filename:
                file_path = f"{f.filename}:{f.lineno}"
                break
    if file_path:
        source = f"{file_path}:{name}"
    return hashlib.sha256(source.encode()).hexdigest()[:16]


def _validate_name(name: str) -> None:
    if not _NAME_PATTERN.match(name):
        raise SpecRegistrationError(
            f"Invalid spec name '{name}': must be 1-100 chars, alphanumeric/hyphens/underscores"
        )


def define_eval(
    name_or_config: Any = None,
    executor: Callable[..., Any] | None = None,
    *,
    name: str | None = None,
    options: SpecOptions | None = None,
    description: str | None = None,
    suite: str | None = None,
    tags: list[str] | None = None,
    timeout_ms: int = 30_000,
) -> EvalSpec | None:
    """Register an eval spec with the active runtime.

    Can be called as::

        # Positional style
        define_eval("my-test", my_executor)

        # Config style
        define_eval(SpecConfig(name="my-test", executor=my_executor))

        # Decorator style
        @define_eval(name="my-test")
        async def my_test(ctx):
            ...
    """
    if isinstance(name_or_config, SpecConfig):
        cfg = name_or_config
        spec_name = cfg.name
        spec_executor = cfg.executor
        spec_options = cfg.options
        spec_desc = cfg.description
        spec_suite = cfg.suite
    elif isinstance(name_or_config, str):
        spec_name = name_or_config
        spec_executor = executor
        spec_options = options or SpecOptions(timeout_ms=timeout_ms, tags=tags or [])
        spec_desc = description
        spec_suite = suite
    elif name_or_config is None and name is not None:
        spec_name = name
        spec_executor = executor
        spec_options = options or SpecOptions(timeout_ms=timeout_ms, tags=tags or [])
        spec_desc = description
        spec_suite = suite
    elif callable(name_or_config) and name is not None:
        spec_name = name
        spec_executor = name_or_config
        spec_options = options or SpecOptions(timeout_ms=timeout_ms, tags=tags or [])
        spec_desc = description
        spec_suite = suite
    else:
        # Decorator mode — return a decorator
        def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
            define_eval(
                fn,
                name=name or fn.__name__,
                options=options,
                description=description,
                suite=suite,
                tags=tags,
                timeout_ms=timeout_ms,
            )
            return fn

        if callable(name_or_config):
            return decorator(name_or_config)
        return decorator  # type: ignore[return-value]

    _validate_name(spec_name)
    spec_id = _generate_spec_id(spec_name)

    spec = EvalSpec(
        id=spec_id,
        name=spec_name,
        executor=spec_executor,
        options=spec_options,
        suite=spec_suite,
        description=spec_desc,
    )

    runtime = get_active_runtime()
    if runtime is not None:
        runtime.register(spec)

    return spec


def _define_eval_with_mode(
    mode: Literal["normal", "skip", "only"],
    name_or_config: Any = None,
    executor: Callable[..., Any] | None = None,
    *,
    name: str | None = None,
    options: SpecOptions | None = None,
    description: str | None = None,
    suite: str | None = None,
    tags: list[str] | None = None,
    timeout_ms: int = 30_000,
) -> EvalSpec | None:
    """Internal: register with a specific mode."""
    spec = define_eval(
        name_or_config,
        executor,
        name=name,
        options=options,
        description=description,
        suite=suite,
        tags=tags,
        timeout_ms=timeout_ms,
    )
    if isinstance(spec, EvalSpec):
        spec.mode = mode
    return spec


def define_eval_skip(
    name_or_config: Any = None,
    executor: Callable[..., Any] | None = None,
    **kwargs: Any,
) -> EvalSpec | None:
    """Register a spec but skip it during execution (vitest/jest ``.skip`` convention)."""
    return _define_eval_with_mode("skip", name_or_config, executor, **kwargs)


def define_eval_only(
    name_or_config: Any = None,
    executor: Callable[..., Any] | None = None,
    **kwargs: Any,
) -> EvalSpec | None:
    """Register a spec for exclusive execution (vitest/jest ``.only`` convention)."""
    return _define_eval_with_mode("only", name_or_config, executor, **kwargs)


class _EvalAI:
    """Convenience namespace — ``evalai.test`` is an alias for ``define_eval``."""

    test = staticmethod(define_eval)


evalai = _EvalAI()

# Attach .skip / .only on define_eval to match TS defineEval.skip() / defineEval.only()
define_eval.skip = define_eval_skip  # type: ignore[attr-defined]
define_eval.only = define_eval_only  # type: ignore[attr-defined]


# ── skip/only filtering ──────────────────────────────────────────────


def get_filtered_specs(specs: list[EvalSpec]) -> list[EvalSpec]:
    """Apply skip/only semantics to a list of specs.

    If any spec has ``mode == "only"``, return only those.
    Otherwise, return all specs except those with ``mode == "skip"``.
    """
    only_specs = [s for s in specs if s.mode == "only"]
    if only_specs:
        return only_specs
    return [s for s in specs if s.mode != "skip"]


# ── from_dataset ─────────────────────────────────────────────────────


def _parse_jsonl(content: str) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    for i, line in enumerate(content.splitlines()):
        line = line.strip()
        if not line:
            continue
        try:
            rows.append(json.loads(line))
        except json.JSONDecodeError as exc:
            raise SpecRegistrationError(f"Invalid JSON on line {i + 1} of dataset: {exc}") from exc
    return rows


def _parse_csv(content: str) -> list[dict[str, Any]]:
    reader = csv.DictReader(io.StringIO(content))
    return [dict(row) for row in reader]


def from_dataset(
    name: str,
    dataset_path: str,
    executor: Callable[..., Any],
    options: SpecOptions | None = None,
) -> None:
    """Load a JSONL, CSV, or JSON dataset and register one spec per row.

    Each row is passed as ``context.input`` to the executor.

    Example::

        from_dataset("rag-accuracy", "./evals/golden.jsonl", my_executor)
    """
    resolved = os.path.abspath(dataset_path)
    if not os.path.isfile(resolved):
        raise SpecRegistrationError(f"Dataset file not found: {resolved}")

    with open(resolved, encoding="utf-8") as f:
        content = f.read()

    ext = os.path.splitext(resolved)[1].lower()
    if ext in (".jsonl", ".ndjson"):
        rows = _parse_jsonl(content)
    elif ext == ".csv":
        rows = _parse_csv(content)
    elif ext == ".json":
        parsed = json.loads(content)
        rows = parsed if isinstance(parsed, list) else [parsed]
    else:
        raise SpecRegistrationError(f"Unsupported dataset format: {ext}. Use .jsonl, .ndjson, .csv, or .json")

    if not rows:
        raise SpecRegistrationError(f"Dataset is empty: {resolved}")

    for i, row in enumerate(rows):
        row_name = f"{name} - row {i + 1}"

        def _make_wrapper(r: dict[str, Any]) -> Callable[..., Any]:
            async def wrapper(ctx: EvalContext) -> EvalResult:
                ctx.input = r
                return await executor(ctx)

            return wrapper

        row_options = SpecOptions(
            timeout_ms=options.timeout_ms if options else 30_000,
            tags=list(options.tags) if options and options.tags else [],
            metadata={
                **(options.metadata or {} if options else {}),
                "dataset_path": resolved,
                "dataset_row": i + 1,
            },
        )
        define_eval(row_name, _make_wrapper(row), options=row_options)


def define_suite(name: str, specs: list[Callable[[], None]]) -> None:
    """Group multiple define_eval calls into a named suite."""
    for spec_fn in specs:
        spec_fn()


def create_result(
    *,
    passed: bool,
    score: float = 0.0,
    assertions: list[Any] | None = None,
    metadata: dict[str, Any] | None = None,
    error: str | None = None,
    output: str | None = None,
    tokens: int | None = None,
    duration_ms: float | None = None,
) -> EvalResult:
    """Create an evaluation result."""
    clamped_score = max(0.0, min(100.0, score))
    return EvalResult(
        passed=passed,
        score=clamped_score,
        assertions=assertions or [],
        metadata=metadata or {},
        error=error,
        status="passed" if passed else ("error" if error else "failed"),
        output=output,
        tokens=tokens,
        duration_ms=duration_ms or 0.0,
    )
